File: OpenACC/calor2DAcc/ecCalor2D_tdma_parallel.f90
PROGRAM Calor2D
!
use constantes
implicit none !(type, external)
! external :: sgesv
integer :: ii, jj, kk
!
! Variables de tamaño para la placa
double precision :: a, b, deltax, deltay
!
! Parámetros físicos del problema
double precision :: cond_ter, temp_ini, temp_fin
double precision :: flux_aba, flux_arr, alpha
!
! Arreglos para incógnitas, término fuente y posición
double precision, dimension(mi,nj) :: temper,temp_ant
double precision, dimension(mi)    :: sourcex, xx
double precision, dimension(nj)    :: sourcey, yy
double precision, dimension(mi,nj) :: resultx
double precision, dimension(nj,mi) :: resulty,tempy
!
! Matriz a invertir
double precision, dimension(mi,nj) :: AI,AD
double precision, dimension(mi,nj) :: AC
double precision, dimension(nj,mi) :: BI,BD
double precision, dimension(nj,mi) :: BC
!
! Se crea la malla 2D
!
a = 1.d0
b = 5.d-1
do ii = 1, mi
    xx(ii) = a*dfloat(ii-1)/dfloat(mi-1)
!     print*, ii, xx(ii)
end do
do jj = 1, nj
    yy(jj) = b*dfloat(jj-1)/dfloat(nj-1)
end do
deltax = 1.d0/dfloat(mi-1)
deltay = 1.d0/dfloat(nj-1)
!
! Se definen los parámetros físicos del problema
!
cond_ter = 1.d0
temp_ini = 308.d0
temp_fin = 298.d0
flux_aba = 5.0d0
flux_arr = 5.0d0
alpha    = 0.5d0 ! par\'ametro de relajaci\'on
!
! Inicialización de los arreglos a utilizar
!
AI = 0.d0; AC = 0.d0; AD = 0.d0
BI = 0.d0; BC = 0.d0; BD = 0.d0
temper = (temp_fin+temp_ini)/2.d0
temp_ant = temper
resultx = 0.0d0
resulty = 0.0d0
!
! Abrimos la región de datos paralela
!
!$acc data copy( temper(1:mi,1:nj) ) &
!$acc & copyin( deltax,deltay,temp_ant(1:mi,1:nj),cond_ter,temp_ini,&
!$acc & temp_fin,flux_aba,flux_arr,alpha,resultx(1:mi,1:nj),resulty(1:nj,1:mi),&
!$acc & AI(1:mi,1:nj),AC(1:mi,1:nj),AD(1:mi,1:nj),&
!$acc & BI(1:nj,1:mi),BC(1:nj,1:mi),BD(1:nj,1:mi),tempy(1:nj,1:mi) )
!
! Bucle de pseudotiempo
! 
do kk = 1, 500
   !
   ! Inicia el ciclo que recorre la coordenada y resolviendo
   ! problemas 1D en la dirección de x
   
   !$acc parallel loop
   do jj = 2, nj-1
      !
      ! Ensamblamos matrices en direcci\'on x
      !
      call ensambla_tdmax(AI,AC,AD,resultx,deltax,deltay,temp_ant,cond_ter,temp_ini,temp_fin,jj)
      !
      ! Llamamos al TDMA
      call tri(AI(1:mi,jj),AC(1:mi,jj),AD(1:mi,jj),resultx(1:mi,jj),temper(1:mi,jj),mi)
      
    end do
    !
    ! Inicia el ciclo que recorre la coordenada x resolviendo
    ! problemas 1D en la dirección de y
    !
    !$acc parallel loop
    do ii = 2, mi-1
       !
       ! Ensamblamos matrices tridiagonales en direcci\'on y
       !
       call ensambla_tdmay(BI,BC,BD,resulty,deltax,deltay,temper,cond_ter,flux_aba,flux_arr,ii)
       !
       ! Llamamos al TDMA
       !
       call tri(BI(1:nj,ii),BC(1:nj,ii),BD(1:nj,ii),resulty(1:nj,ii),tempy(1:nj,ii),nj)
       do jj =1, nj
          temper(ii,jj) = tempy(jj,ii)
       end do
       
    end do
    !
    ! se aplica un esquema de relajaci\'on fija
    !
    temper = alpha*temper+(1.d0-alpha)*temp_ant
    !
    ! se actualiza la temperatura de la iteraci\'on anterior
    !
    temp_ant = temper
    !
    !print*, "Residuo: ", kk, maxval(temper-temp_ant)
    !     temp_ant = temper
    !
 end do
 !
 ! Cerramos la región de datos paralela
 !
 !$acc end data
 !
 !! Escritura de resultados
 !
 do ii = 1, mi
    do jj = 1, nj
       write(101,*) xx(ii), yy(jj), temper(ii,jj)
    end do
    write(101,*)
 end do
 !
 ! Obtenemos la diferencia entre la solución analítica y la solución numérica, así como el error porcentual
 !do ii =2, mi-1
 !    do jj = 2, nj-1
 !        !
 !        write(102,*) xx(ii), yy(jj), temper(ii,jj)-(-xx(ii)+308), (temper(ii,jj)-(-xx(ii)+308))/(-xx(ii)+308)
 !    end do
 !        write(102,*)
 !end do
 !
 !
 ! Obtenemos la diferencia entre la solución analítica y la solución numérica, así como el error porcentual
 ! do ii =2, mi-1
 !     do jj = 2, nj-1
 !         !
 !         write(102,*) xx(ii), yy(jj), temper(ii,jj)-(-xx(ii)+308), (temper(ii,jj)-(-xx(ii)+308))/(temp_fin-temp_ini)
 !     end do
 !         write(102,*)
 ! end do
 !
END PROGRAM Calor2D

File: CUDA/ec_calor_2d_tdma_parallel.c
#include <stdio.h>
#include <stdlib.h>

#include "constantes.h"
#include "herramientas.h"

int main(int argc, char const *argv[])
{
    int ii, jj, kk;
    
    // Variables de tamano de la placa
    double a, b, deltax, deltay;

    // Parametros fisicos del problema
    double cond_ter, temp_ini, temp_fin;
    double flux_aba, flux_arr, alpha;

    // Arreglos para incognitas, termino fuente y posicion
    double temper[mi][nj], temp_ant[mi][nj];
    double sourcex[mi], xx[mi];
    double sourcey[nj], yy[nj];
    double resultx[mi][nj], tempx[mi][nj];
    double resulty[nj][mi], tempy[nj][mi];

    //Matriz a invertir
    double AI[mi][nj], AD[mi][nj], AC[mi][nj];
    double BI[nj][mi], BD[nj][mi], BC[nj][mi];
    /*
    * Se crea la malla 2D
    */
    a = 10.0;
    b = 5.0;

    for (ii = 0; ii <= mi-1; ii++)
    {
        double dii1 = (double)ii;
        double dmi1 = (double)(mi - 1);
        xx[ii] = a * dii1 / dmi1;
    }
    for (jj = 0; jj <= nj-1; jj++)
    {
        double djj1 = (double)jj;
        double dnj1 = (double)(nj -1);
        yy[jj] = b * djj1 / dnj1;
    }
    double dmi1 = (double)(mi - 1);
    deltax = 1.0 / dmi1;
    double dnj1 = (double)(nj - 1);
    deltay = 1.0 / dnj1;
    /*
    * Se definen los parametros fisicos del problema
    */
    cond_ter = 100.0;
    temp_ini = 308.0;
    temp_fin = 298.0;
    flux_aba = 5.0;
    flux_arr = 5.0;
    alpha    = 0.5; //parametro de relajacion
    /*
    * Inicializacion de los arreglos a utilizar
    */
    inicializar_matriz(AI, mi, nj, 0.0);
    inicializar_matriz(AC, mi, nj, 0.0);
    inicializar_matriz(AD, mi, nj, 0.0);
    inicializar_matriz(BI, nj, mi, 0.0);
    inicializar_matriz(BC, nj, mi, 0.0);
    inicializar_matriz(BD, nj, mi, 0.0);
    
    double valor = (temp_fin+temp_ini)/2.0;
    inicializar_matriz(temper, mi, nj, valor);
    inicializar_matriz(temp_ant, mi, nj, valor);

    inicializar_matriz(resultx, mi, nj, 0.0);
    inicializar_matriz(resulty, nj, mi, 0.0);
    inicializar_matriz(tempx, mi, nj, 0.0);
    inicializar_matriz(tempy, nj, mi, 0.0);
    /*
    * Bucle de pseudotiempo
    */
    for (kk = 0; kk < 30; kk++)
    {
        /*
        * Inicia el ciclo que recorre la coordenada y resolviendo problemas 1D en la direccion de x
        */
       for (jj = 1; jj < nj-1; jj++)
       {
            /*
            * Ensamblando matrices en direccion x
            */
            ensambla_tdmax(AI,AC,AD,resultx,deltax,deltay,temp_ant,cond_ter,temp_ini,temp_fin,jj);
       }
       /*
       * Llamamos al resolvedor
       */
        for (jj = 1; jj < nj-1; jj++)
        {
            // TODO: Crear funcion para tomar columnas de una matriz como vectores
            // * Llamar a la funcion tri
            for (ii = 0; ii < mi; ii++)
            {
                temper[ii][jj] = resultx[ii][jj];
            }
            
        }
        /*
        * Inicia el ciclo que recorre la coordenada x resolviendo problemas 1D en la direccion de y
        */
        for (ii = 1; ii < mi-1; ii++)
        {
            /*
            * Ensamblamos matrices tridiagonales en direccion y
            */
           ensambla_tdmay(BI,BC,BD,resulty,deltax,deltay,temper,cond_ter,flux_aba,flux_arr,ii);
        }
        /*
        * Llamamos al resolvedor
        */
        for (ii = 1; ii < mi-1; ii++)
        {
            // TODO: Crear funcion para tomar columnas de una matriz como vectores
            // * Llamar a la funcion tri
            for (jj = 0; jj < nj; jj++)
            {
                temper[ii][jj] = resulty[jj][ii];
            }
        }
        /*
        * Se actualiza la temperatura de la iteracion anterior
        */
        for (ii = 0; ii < mi; ii++)
        {
            for (jj = 0; jj < nj; jj++)
            {
                temp_ant[ii][jj] = temper[ii][jj];
            }
        }
        
    }
    /*
    * Escritura de resultados
    */
   // Abrir el archivo para escribir
    FILE *file = fopen("fort.101", "w");

    if (file != NULL) {
        // Utilizar un bucle para imprimir y guardar los datos
        for (ii = 0; ii < mi; ii++) {
            for (jj = 0; jj < nj; jj++)
            {
                fprintf(file, "%f %f %f\n", xx[ii], yy[ii], temper[ii][jj]);
            }
        }
        // Cerrar el archivo después de escribir
        fclose(file);
    } else {
        printf("Error al abrir el archivo.\n");
    }
    return 0;
}
